File: server/services/ai/router.ts
/**
 * Intelligent model router - selects the best AI model for each request
 * Supports 7+ providers with automatic selection based on task requirements
 */
import type { AIRequest, ModelProvider, ModelInfo } from '../../types/index.js'
import { OpenAIService } from './openai.js'
import { AnthropicService } from './anthropic.js'
import { GoogleService } from './google.js'
import { PerplexityService } from './perplexity.js'
import { DeepSeekService } from './deepseek.js'
import { MoonshotService } from './moonshot.js'
import { MistralService } from './mistral.js'
import type { AIService } from './base.js'
import { config } from '../../config/index.js'

export class ModelRouter {
  private services: Map<ModelProvider, AIService>
  
  constructor() {
    this.services = new Map()
    
    // Initialize available services based on configured API keys
    const providerConfigs: Array<{ key: ModelProvider; apiKey: string; ServiceClass: new () => AIService }> = [
      { key: 'openai', apiKey: config.ai.openai.apiKey, ServiceClass: OpenAIService },
      { key: 'anthropic', apiKey: config.ai.anthropic.apiKey, ServiceClass: AnthropicService },
      { key: 'google', apiKey: config.ai.google.apiKey, ServiceClass: GoogleService },
      { key: 'perplexity', apiKey: config.ai.perplexity.apiKey, ServiceClass: PerplexityService },
      { key: 'deepseek', apiKey: config.ai.deepseek.apiKey, ServiceClass: DeepSeekService },
      { key: 'moonshot', apiKey: config.ai.moonshot.apiKey, ServiceClass: MoonshotService },
      { key: 'mistral', apiKey: config.ai.mistral.apiKey, ServiceClass: MistralService },
    ]
    
    for (const { key, apiKey, ServiceClass } of providerConfigs) {
      try {
        if (apiKey) {
          this.services.set(key, new ServiceClass())
          console.log(`  ✓ ${key} provider initialized`)
        }
      } catch (error) {
        console.warn(`  ✗ ${key} service not available:`, error instanceof Error ? error.message : error)
      }
    }
    
    console.log(`  → ${this.services.size} AI provider(s) active`)
  }
  
  /**
   * Intelligently selects the best model for a given request
   */
  selectBestModel(request: AIRequest): ModelInfo {
    // Check for specific provider request (e.g. model="perplexity-..." or "deepseek-...")
    const requestedProvider = this.detectProvider(request.model)
    
    // If specific model requested and available, use it
    if (requestedProvider && requestedProvider !== 'auto' && this.services.has(requestedProvider)) {
      const service = this.services.get(requestedProvider)!
      return {
        provider: requestedProvider,
        model: request.model || this.getDefaultModel(requestedProvider),
        capabilities: service.capabilities,
      }
    }
    
    // Auto-select based on request characteristics
    const contextLength = this.estimateContextLength(request)
    const requiresVision = this.requiresVision(request)
    const requiresSpeed = this.requiresSpeed(request)
    const requiresQuality = this.requiresQuality(request)
    const requiresSearch = this.requiresSearch(request)
    const requiresCoding = this.requiresCoding(request)
    
    // Score each available service
    const scores: Array<{ provider: ModelProvider; service: AIService; score: number }> = []
    
    for (const [provider, service] of this.services.entries()) {
      const score = this.scoreService(
        provider, service, contextLength,
        requiresVision, requiresSpeed, requiresQuality,
        requiresSearch, requiresCoding
      )
      scores.push({ provider, service, score })
    }
    
    if (scores.length === 0) {
      throw new Error('No AI services available. Please configure at least one API key in .env')
    }
    
    // Sort by score and select best
    scores.sort((a, b) => b.score - a.score)
    const best = scores[0]
    
    return {
      provider: best.provider,
      model: this.getDefaultModel(best.provider),
      capabilities: best.service.capabilities,
    }
  }
  
  private detectProvider(model?: string): ModelProvider | undefined {
    if (!model) return undefined
    
    const modelLower = model.toLowerCase()
    
    if (modelLower.includes('gpt') || modelLower.includes('openai')) return 'openai'
    if (modelLower.includes('claude') || modelLower.includes('anthropic')) return 'anthropic'
    if (modelLower.includes('gemini') || modelLower.includes('google')) return 'google'
    if (modelLower.includes('sonar') || modelLower.includes('perplexity')) return 'perplexity'
    if (modelLower.includes('deepseek')) return 'deepseek'
    if (modelLower.includes('moonshot') || modelLower.includes('kimi')) return 'moonshot'
    if (modelLower.includes('mistral') || modelLower.includes('mixtral')) return 'mistral'
    
    return undefined
  }
  
  private scoreService(
    provider: ModelProvider,
    service: AIService,
    contextLength: number,
    requiresVision: boolean,
    requiresSpeed: boolean,
    requiresQuality: boolean,
    requiresSearch: boolean,
    requiresCoding: boolean,
  ): number {
    const caps = service.capabilities
    let score = 0
    
    // Context length match
    if (caps.maxContextLength >= contextLength) {
      score += 10
    } else {
      score -= 20
    }
    
    // Vision support
    if (requiresVision && caps.supportsVision) {
      score += 5
    } else if (requiresVision && !caps.supportsVision) {
      score -= 10
    }
    
    // Speed preference
    if (requiresSpeed) {
      if (caps.speed === 'fast') score += 5
      else if (caps.speed === 'medium') score += 2
    }
    
    // Quality preference
    if (requiresQuality) {
      if (caps.quality === 'high') score += 5
      else if (caps.quality === 'medium') score += 2
    }
    
    // Search-enhanced tasks -> prefer Perplexity
    if (requiresSearch && provider === 'perplexity') {
      score += 15
    }
    
    // Coding tasks -> prefer DeepSeek (code-specialized) or Anthropic (excellent at code)
    if (requiresCoding) {
      if (provider === 'deepseek') score += 10
      if (provider === 'anthropic') score += 8
    }
    
    // Cost efficiency
    const avgCost = (caps.costPer1kTokens.input + caps.costPer1kTokens.output) / 2
    if (avgCost > 0) {
      score += (0.01 / avgCost) * 2
    }
    
    return score
  }
  
  private estimateContextLength(request: AIRequest): number {
    let length = 0
    
    for (const msg of request.messages) {
      length += Math.ceil(msg.content.length / 4)
    }
    
    if (request.context?.files) {
      for (const file of request.context.files) {
        length += Math.ceil(file.content.length / 4)
      }
    }
    
    return length
  }
  
  private requiresVision(request: AIRequest): boolean {
    const content = request.messages.map(m => m.content.toLowerCase()).join(' ')
    return content.includes('image') || 
           content.includes('screenshot') || 
           content.includes('visual') ||
           content.includes('design') ||
           (request.context?.files?.some(f => 
             f.path.match(/\.(png|jpg|jpeg|gif|svg)$/i)
           ) ?? false)
  }
  
  private requiresSpeed(request: AIRequest): boolean {
    const content = request.messages.map(m => m.content.toLowerCase()).join(' ')
    const simpleTasks = ['explain', 'summarize', 'translate', 'format', 'refactor simple']
    return simpleTasks.some(task => content.includes(task))
  }
  
  private requiresQuality(request: AIRequest): boolean {
    const content = request.messages.map(m => m.content.toLowerCase()).join(' ')
    const complexTasks = ['architecture', 'design system', 'complex', 'critical', 'production', 'security audit']
    return complexTasks.some(task => content.includes(task))
  }
  
  private requiresSearch(request: AIRequest): boolean {
    const content = request.messages.map(m => m.content.toLowerCase()).join(' ')
    const searchTerms = [
      'search', 'find online', 'latest', 'current', 'today', 'news',
      'documentation', 'look up', 'what is the latest', 'recent',
      'up to date', 'real-time', 'web search', 'browse',
    ]
    return searchTerms.some(term => content.includes(term))
  }
  
  private requiresCoding(request: AIRequest): boolean {
    const content = request.messages.map(m => m.content.toLowerCase()).join(' ')
    const codeTerms = [
      'code', 'function', 'class', 'implement', 'debug', 'fix bug',
      'refactor', 'optimize', 'typescript', 'javascript', 'python',
      'rust', 'algorithm', 'api', 'endpoint', 'component',
    ]
    return codeTerms.some(term => content.includes(term))
  }
  
  private getDefaultModel(provider: ModelProvider): string {
    const defaults: Record<string, string> = {
      openai: 'gpt-4o',
      anthropic: 'claude-opus-4-6',
      google: 'gemini-2.0-flash',
      perplexity: 'llama-3.1-sonar-large-128k-online',
      deepseek: 'deepseek-chat',
      moonshot: 'moonshot-v1-128k',
      mistral: 'mistral-large-2512',
      auto: 'gpt-4o',
    }
    return defaults[provider] || 'gpt-4o'
  }
  
  getService(provider: ModelProvider): AIService | undefined {
    return this.services.get(provider)
  }
  
  getAvailableProviders(): ModelProvider[] {
    return Array.from(this.services.keys())
  }
  
  getProviderCount(): number {
    return this.services.size
  }
}

File: backend/src/services/ai/router.rs
/**
 * Intelligent model router - selects the best AI model for each request
 * This is what makes Bloop superior to KIMI and Claude
 * Supports 15+ AI providers with intelligent selection
 */
use crate::types::{AIRequest, ModelProvider, ModelInfo, ModelCapabilities};
use crate::services::ai::{
    OpenAIService, AnthropicService, GoogleService, MoonshotService,
    DeepSeekService, MistralService, CohereService, PerplexityService,
    XAIService, TogetherService, AnyscaleService, QwenService,
    ZeroOneService, BaiduService
};
use crate::services::ai::base::AIService;
use crate::config::Config;
use std::sync::Arc;

pub struct ModelRouter {
    openai: Option<Arc<OpenAIService>>,
    anthropic: Option<Arc<AnthropicService>>,
    google: Option<Arc<GoogleService>>,
    moonshot: Option<Arc<MoonshotService>>,
    deepseek: Option<Arc<DeepSeekService>>,
    mistral: Option<Arc<MistralService>>,
    cohere: Option<Arc<CohereService>>,
    perplexity: Option<Arc<PerplexityService>>,
    xai: Option<Arc<XAIService>>,
    together: Option<Arc<TogetherService>>,
    anyscale: Option<Arc<AnyscaleService>>,
    qwen: Option<Arc<QwenService>>,
    zeroone: Option<Arc<ZeroOneService>>,
    baidu: Option<Arc<BaiduService>>,
}

/// Helper enum to hold different service types
#[derive(Clone)]
pub enum AIServiceEnum {
    OpenAI(Arc<OpenAIService>),
    Anthropic(Arc<AnthropicService>),
    Google(Arc<GoogleService>),
    Moonshot(Arc<MoonshotService>),
    DeepSeek(Arc<DeepSeekService>),
    Mistral(Arc<MistralService>),
    Cohere(Arc<CohereService>),
    Perplexity(Arc<PerplexityService>),
    XAI(Arc<XAIService>),
    Together(Arc<TogetherService>),
    Anyscale(Arc<AnyscaleService>),
    Qwen(Arc<QwenService>),
    ZeroOne(Arc<ZeroOneService>),
    Baidu(Arc<BaiduService>),
}

impl AIService for AIServiceEnum {
    fn name(&self) -> &str {
        match self {
            AIServiceEnum::OpenAI(s) => s.name(),
            AIServiceEnum::Anthropic(s) => s.name(),
            AIServiceEnum::Google(s) => s.name(),
            AIServiceEnum::Moonshot(s) => s.name(),
            AIServiceEnum::DeepSeek(s) => s.name(),
            AIServiceEnum::Mistral(s) => s.name(),
            AIServiceEnum::Cohere(s) => s.name(),
            AIServiceEnum::Perplexity(s) => s.name(),
            AIServiceEnum::XAI(s) => s.name(),
            AIServiceEnum::Together(s) => s.name(),
            AIServiceEnum::Anyscale(s) => s.name(),
            AIServiceEnum::Qwen(s) => s.name(),
            AIServiceEnum::ZeroOne(s) => s.name(),
            AIServiceEnum::Baidu(s) => s.name(),
        }
    }
    
    fn capabilities(&self) -> &ModelCapabilities {
        match self {
            AIServiceEnum::OpenAI(s) => s.capabilities(),
            AIServiceEnum::Anthropic(s) => s.capabilities(),
            AIServiceEnum::Google(s) => s.capabilities(),
            AIServiceEnum::Moonshot(s) => s.capabilities(),
            AIServiceEnum::DeepSeek(s) => s.capabilities(),
            AIServiceEnum::Mistral(s) => s.capabilities(),
            AIServiceEnum::Cohere(s) => s.capabilities(),
            AIServiceEnum::Perplexity(s) => s.capabilities(),
            AIServiceEnum::XAI(s) => s.capabilities(),
            AIServiceEnum::Together(s) => s.capabilities(),
            AIServiceEnum::Anyscale(s) => s.capabilities(),
            AIServiceEnum::Qwen(s) => s.capabilities(),
            AIServiceEnum::ZeroOne(s) => s.capabilities(),
            AIServiceEnum::Baidu(s) => s.capabilities(),
        }
    }
    
    async fn generate(&self, request: AIRequest) -> anyhow::Result<crate::types::AIResponse> {
        match self {
            AIServiceEnum::OpenAI(s) => s.generate(request).await,
            AIServiceEnum::Anthropic(s) => s.generate(request).await,
            AIServiceEnum::Google(s) => s.generate(request).await,
            AIServiceEnum::Moonshot(s) => s.generate(request).await,
            AIServiceEnum::DeepSeek(s) => s.generate(request).await,
            AIServiceEnum::Mistral(s) => s.generate(request).await,
            AIServiceEnum::Cohere(s) => s.generate(request).await,
            AIServiceEnum::Perplexity(s) => s.generate(request).await,
            AIServiceEnum::XAI(s) => s.generate(request).await,
            AIServiceEnum::Together(s) => s.generate(request).await,
            AIServiceEnum::Anyscale(s) => s.generate(request).await,
            AIServiceEnum::Qwen(s) => s.generate(request).await,
            AIServiceEnum::ZeroOne(s) => s.generate(request).await,
            AIServiceEnum::Baidu(s) => s.generate(request).await,
        }
    }
}

impl ModelRouter {
    pub fn new(config: &Config) -> Self {
        Self {
            openai: if !config.openai_api_key.is_empty() {
                Some(Arc::new(OpenAIService::new(config)))
            } else {
                None
            },
            anthropic: if !config.anthropic_api_key.is_empty() {
                Some(Arc::new(AnthropicService::new(config)))
            } else {
                None
            },
            google: if !config.google_gemini_api_key.is_empty() {
                Some(Arc::new(GoogleService::new(config)))
            } else {
                None
            },
            moonshot: if !config.moonshot_api_key.is_empty() {
                Some(Arc::new(MoonshotService::new(config)))
            } else {
                None
            },
            deepseek: if !config.deepseek_api_key.is_empty() {
                Some(Arc::new(DeepSeekService::new(config)))
            } else {
                None
            },
            mistral: if !config.mistral_api_key.is_empty() {
                Some(Arc::new(MistralService::new(config)))
            } else {
                None
            },
            cohere: if !config.cohere_api_key.is_empty() {
                Some(Arc::new(CohereService::new(config)))
            } else {
                None
            },
            perplexity: if !config.perplexity_api_key.is_empty() {
                Some(Arc::new(PerplexityService::new(config)))
            } else {
                None
            },
            xai: if !config.xai_api_key.is_empty() {
                Some(Arc::new(XAIService::new(config)))
            } else {
                None
            },
            together: if !config.together_api_key.is_empty() {
                Some(Arc::new(TogetherService::new(config)))
            } else {
                None
            },
            anyscale: if !config.anyscale_api_key.is_empty() {
                Some(Arc::new(AnyscaleService::new(config)))
            } else {
                None
            },
            qwen: if !config.qwen_api_key.is_empty() {
                Some(Arc::new(QwenService::new(config)))
            } else {
                None
            },
            zeroone: if !config.zeroone_api_key.is_empty() {
                Some(Arc::new(ZeroOneService::new(config)))
            } else {
                None
            },
            baidu: if !config.baidu_api_key.is_empty() {
                Some(Arc::new(BaiduService::new(config)))
            } else {
                None
            },
        }
    }
    
    /// Intelligently selects the best model for a given request
    /// Considers: context length, cost, speed, quality, task type
    pub fn select_best_model(&self, request: &AIRequest) -> anyhow::Result<ModelInfo> {
        // Check if specific model requested
        if let Some(model_str) = &request.model {
            if let Some(provider) = self.parse_provider_from_model(model_str) {
                if let Some(service) = self.get_service(provider) {
                    return Ok(ModelInfo {
                        provider,
                        model: model_str.clone(),
                        capabilities: service.capabilities().clone(),
                    });
                }
            }
        }
        
        // Auto-select based on request characteristics
        let context_length = self.estimate_context_length(request);
        let requires_vision = self.requires_vision(request);
        let requires_speed = self.requires_speed(request);
        let requires_quality = self.requires_quality(request);
        
        // Score each available service
        let mut scores: Vec<(ModelProvider, f64, ModelCapabilities)> = Vec::new();
        
        if let Some(ref service) = self.openai {
            let score = self.score_service(service.as_ref(), context_length, requires_vision, requires_speed, requires_quality);
            scores.push((ModelProvider::OpenAI, score, service.capabilities().clone()));
        }
        
        if let Some(ref service) = self.anthropic {
            let score = self.score_service(service.as_ref(), context_length, requires_vision, requires_speed, requires_quality);
            scores.push((ModelProvider::Anthropic, score, service.capabilities().clone()));
        }
        
        if let Some(ref service) = self.google {
            let score = self.score_service(service.as_ref(), context_length, requires_vision, requires_speed, requires_quality);
            scores.push((ModelProvider::Google, score, service.capabilities().clone()));
        }
        
        if let Some(ref service) = self.moonshot {
            let score = self.score_service(service.as_ref(), context_length, requires_vision, requires_speed, requires_quality);
            scores.push((ModelProvider::Moonshot, score, service.capabilities().clone()));
        }
        
        if let Some(ref service) = self.deepseek {
            let score = self.score_service(service.as_ref(), context_length, requires_vision, requires_speed, requires_quality);
            scores.push((ModelProvider::DeepSeek, score, service.capabilities().clone()));
        }
        
        if let Some(ref service) = self.mistral {
            let score = self.score_service(service.as_ref(), context_length, requires_vision, requires_speed, requires_quality);
            scores.push((ModelProvider::Mistral, score, service.capabilities().clone()));
        }
        
        if let Some(ref service) = self.cohere {
            let score = self.score_service(service.as_ref(), context_length, requires_vision, requires_speed, requires_quality);
            scores.push((ModelProvider::Cohere, score, service.capabilities().clone()));
        }
        
        if let Some(ref service) = self.perplexity {
            let score = self.score_service(service.as_ref(), context_length, requires_vision, requires_speed, requires_quality);
            scores.push((ModelProvider::Perplexity, score, service.capabilities().clone()));
        }
        
        if let Some(ref service) = self.xai {
            let score = self.score_service(service.as_ref(), context_length, requires_vision, requires_speed, requires_quality);
            scores.push((ModelProvider::XAI, score, service.capabilities().clone()));
        }
        
        if let Some(ref service) = self.together {
            let score = self.score_service(service.as_ref(), context_length, requires_vision, requires_speed, requires_quality);
            scores.push((ModelProvider::Together, score, service.capabilities().clone()));
        }
        
        if let Some(ref service) = self.anyscale {
            let score = self.score_service(service.as_ref(), context_length, requires_vision, requires_speed, requires_quality);
            scores.push((ModelProvider::Anyscale, score, service.capabilities().clone()));
        }
        
        if let Some(ref service) = self.qwen {
            let score = self.score_service(service.as_ref(), context_length, requires_vision, requires_speed, requires_quality);
            scores.push((ModelProvider::Qwen, score, service.capabilities().clone()));
        }
        
        if let Some(ref service) = self.zeroone {
            let score = self.score_service(service.as_ref(), context_length, requires_vision, requires_speed, requires_quality);
            scores.push((ModelProvider::ZeroOne, score, service.capabilities().clone()));
        }
        
        if let Some(ref service) = self.baidu {
            let score = self.score_service(service.as_ref(), context_length, requires_vision, requires_speed, requires_quality);
            scores.push((ModelProvider::Baidu, score, service.capabilities().clone()));
        }
        
        if scores.is_empty() {
            return Err(anyhow::anyhow!("No AI services available"));
        }
        
        // Sort by score and select best
        scores.sort_by(|a, b| b.1.partial_cmp(&a.1).unwrap_or(std::cmp::Ordering::Equal));
        let (provider, _, capabilities) = &scores[0];
        
        Ok(ModelInfo {
            provider: provider.clone(),
            model: self.get_default_model(provider),
            capabilities: capabilities.clone(),
        })
    }
    
    fn score_service(
        &self,
        service: &dyn AIService,
        context_length: u32,
        requires_vision: bool,
        requires_speed: bool,
        requires_quality: bool,
    ) -> f64 {
        let caps = service.capabilities();
        let mut score = 0.0;
        
        // Context length match (higher is better)
        if caps.max_context_length >= context_length {
            score += 10.0;
        } else {
            score -= 20.0; // Penalty for insufficient context
        }
        
        // Vision support
        if requires_vision && caps.supports_vision {
            score += 5.0;
        }
        
        // Speed preference
        if requires_speed {
            match caps.speed {
                crate::types::Speed::Fast => score += 5.0,
                crate::types::Speed::Medium => score += 2.0,
                crate::types::Speed::Slow => {},
            }
        }
        
        // Quality preference
        if requires_quality {
            match caps.quality {
                crate::types::Quality::High => score += 5.0,
                crate::types::Quality::Medium => score += 2.0,
                crate::types::Quality::Low => {},
            }
        }
        
        // Cost efficiency (lower cost = higher score)
        let avg_cost = (caps.cost_per_1k_tokens.input + caps.cost_per_1k_tokens.output) / 2.0;
        score += (0.01 / avg_cost) * 2.0;
        
        score
    }
    
    fn estimate_context_length(&self, request: &AIRequest) -> u32 {
        let mut length = 0u32;
        
        // Messages
        for msg in &request.messages {
            length += (msg.content.len() as f32 / 4.0).ceil() as u32;
        }
        
        // Context files
        if let Some(ref context) = request.context {
            if let Some(ref files) = context.files {
                for file in files {
                    length += (file.content.len() as f32 / 4.0).ceil() as u32;
                }
            }
        }
        
        length
    }
    
    fn requires_vision(&self, request: &AIRequest) -> bool {
        let content = request.messages
            .iter()
            .map(|m| m.content.to_lowercase())
            .collect::<String>();
        
        content.contains("image") || 
        content.contains("screenshot") || 
        content.contains("visual") ||
        content.contains("design") ||
        request.context.as_ref()
            .and_then(|c| c.files.as_ref())
            .map(|f| f.iter().any(|file| {
                file.path.ends_with(".png") || 
                file.path.ends_with(".jpg") || 
                file.path.ends_with(".jpeg") ||
                file.path.ends_with(".gif") ||
                file.path.ends_with(".svg")
            }))
            .unwrap_or(false)
    }
    
    fn requires_speed(&self, request: &AIRequest) -> bool {
        let content = request.messages
            .iter()
            .map(|m| m.content.to_lowercase())
            .collect::<String>();
        
        let simple_tasks = ["explain", "summarize", "translate", "format", "refactor simple"];
        simple_tasks.iter().any(|task| content.contains(task))
    }
    
    fn requires_quality(&self, request: &AIRequest) -> bool {
        let content = request.messages
            .iter()
            .map(|m| m.content.to_lowercase())
            .collect::<String>();
        
        let complex_tasks = ["architecture", "design", "complex", "critical", "production", "security"];
        complex_tasks.iter().any(|task| content.contains(task))
    }
    
    fn parse_provider_from_model(&self, model: &str) -> Option<ModelProvider> {
        let model_lower = model.to_lowercase();
        if model_lower.starts_with("gpt") || model_lower.starts_with("openai") {
            Some(ModelProvider::OpenAI)
        } else if model_lower.starts_with("claude") || model_lower.starts_with("anthropic") {
            Some(ModelProvider::Anthropic)
        } else if model_lower.starts_with("gemini") || model_lower.starts_with("google") {
            Some(ModelProvider::Google)
        } else if model_lower.starts_with("kimi") || model_lower.starts_with("moonshot") {
            Some(ModelProvider::Moonshot)
        } else if model_lower.starts_with("deepseek") {
            Some(ModelProvider::DeepSeek)
        } else if model_lower.starts_with("mistral") {
            Some(ModelProvider::Mistral)
        } else if model_lower.starts_with("cohere") || model_lower.starts_with("command") {
            Some(ModelProvider::Cohere)
        } else if model_lower.starts_with("perplexity") || model_lower.starts_with("sonar") {
            Some(ModelProvider::Perplexity)
        } else if model_lower.starts_with("grok") || model_lower.starts_with("xai") {
            Some(ModelProvider::XAI)
        } else if model_lower.starts_with("together") || model_lower.contains("llama") || model_lower.contains("mixtral") {
            Some(ModelProvider::Together)
        } else if model_lower.starts_with("anyscale") {
            Some(ModelProvider::Anyscale)
        } else if model_lower.starts_with("qwen") {
            Some(ModelProvider::Qwen)
        } else if model_lower.starts_with("yi") || model_lower.starts_with("zeroone") {
            Some(ModelProvider::ZeroOne)
        } else if model_lower.starts_with("ernie") || model_lower.starts_with("baidu") {
            Some(ModelProvider::Baidu)
        } else {
            None
        }
    }
    
    fn get_default_model(&self, provider: &ModelProvider) -> String {
        match provider {
            ModelProvider::OpenAI => "gpt-4-turbo-preview".to_string(),
            ModelProvider::Anthropic => "claude-3-5-sonnet-20241022".to_string(),
            ModelProvider::Google => "gemini-1.5-pro".to_string(),
            ModelProvider::Moonshot => "kimi-k2.5".to_string(),
            ModelProvider::DeepSeek => "deepseek-chat".to_string(),
            ModelProvider::Mistral => "mistral-large-latest".to_string(),
            ModelProvider::Cohere => "command-r-plus".to_string(),
            ModelProvider::Perplexity => "llama-3.1-sonar-large-128k-online".to_string(),
            ModelProvider::XAI => "grok-beta".to_string(),
            ModelProvider::Together => "meta-llama/Meta-Llama-3-70B-Instruct-Turbo".to_string(),
            ModelProvider::Anyscale => "meta-llama/Meta-Llama-3.1-405B-Instruct".to_string(),
            ModelProvider::Qwen => "qwen-plus".to_string(),
            ModelProvider::ZeroOne => "yi-1.5-34b-chat".to_string(),
            ModelProvider::Baidu => "ernie-4.0-8k".to_string(),
            ModelProvider::Auto => "gpt-4-turbo-preview".to_string(),
        }
    }
    
    pub fn get_service(&self, provider: ModelProvider) -> Option<AIServiceEnum> {
        match provider {
            ModelProvider::OpenAI => self.openai.as_ref().map(|s| AIServiceEnum::OpenAI(s.clone())),
            ModelProvider::Anthropic => self.anthropic.as_ref().map(|s| AIServiceEnum::Anthropic(s.clone())),
            ModelProvider::Google => self.google.as_ref().map(|s| AIServiceEnum::Google(s.clone())),
            ModelProvider::Moonshot => self.moonshot.as_ref().map(|s| AIServiceEnum::Moonshot(s.clone())),
            ModelProvider::DeepSeek => self.deepseek.as_ref().map(|s| AIServiceEnum::DeepSeek(s.clone())),
            ModelProvider::Mistral => self.mistral.as_ref().map(|s| AIServiceEnum::Mistral(s.clone())),
            ModelProvider::Cohere => self.cohere.as_ref().map(|s| AIServiceEnum::Cohere(s.clone())),
            ModelProvider::Perplexity => self.perplexity.as_ref().map(|s| AIServiceEnum::Perplexity(s.clone())),
            ModelProvider::XAI => self.xai.as_ref().map(|s| AIServiceEnum::XAI(s.clone())),
            ModelProvider::Together => self.together.as_ref().map(|s| AIServiceEnum::Together(s.clone())),
            ModelProvider::Anyscale => self.anyscale.as_ref().map(|s| AIServiceEnum::Anyscale(s.clone())),
            ModelProvider::Qwen => self.qwen.as_ref().map(|s| AIServiceEnum::Qwen(s.clone())),
            ModelProvider::ZeroOne => self.zeroone.as_ref().map(|s| AIServiceEnum::ZeroOne(s.clone())),
            ModelProvider::Baidu => self.baidu.as_ref().map(|s| AIServiceEnum::Baidu(s.clone())),
            ModelProvider::Auto => None,
        }
    }
}
